### PR TITLE
salt-cloud/libvirt: enhance libvirt provider

### LINKED ISSUE #52726: Salt-Cloud libvirt provider lacks cpu, memory and networking options
### Description of Issue/Question
Hello, I'm using the Salt-Cloud 'libvirt' provider. Compared to the 'vmware' provider, there are missing options that I need.

Taken from https://docs.saltstack.com/en/latest/topics/cloud/vmware.html:
- num_cpus
- memory
- devices / network

Currently, in the libvirt provider, all above options are inherited from the source virtual machine, there is no way to customize them with salt-cloud.

I am currently developing a patch to add these features.

### Setup
CentOS 7.6, basic setup based on the tutorial:
https://docs.saltstack.com/en/latest/topics/cloud/libvirt.html#provider-configuration

Provider:
```
local-kvm:
  driver: libvirt
  url: qemu:///system
  validate_xml: no
```

Profile:
```
centos7:
  provider: local-kvm
  base_domain: base-centos7-64
  ip_source: qemu-agent
  ssh_username: root
  password: my-very-secret-password
  clone_strategy: quick
  minion:
    master: 192.168.0.35
    master_port: 4506
  num_cpus: 2
  memory: 2048MB
  devices:
    network:
      eth0:
        type: bridge
        mac: '52:54:00:00:00:01'
        source: br0
        model: virtio
      eth1:
        type: network
        mac: '52:54:00:00:00:02'
        source: default
        model: virtio
```

### Steps to Reproduce Issue
`# salt-cloud --log-level=debug -p centos7 my-centos7-clone`

Inspect resulting VM using any libvirt client: you will see that options 'num_cpus', 'memory' and everything under 'devices' are ignored.

### Versions Report
Salt Version:
           Salt: 2019.2.0
 
Dependency Versions:
           cffi: 1.6.0
       cherrypy: Not Installed
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.31.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Apr  9 2019, 14:30:50)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7.6.1810 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-957.10.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.6.1810 Core

### PR DESCRIPTION
**Pull request by f-andriot**
… 'devices/network'

Fixes #52726

### What does this PR do?
Allow salt-cloud libvirt provider to customize 'num_cpus', 'memory' and 'devices/network' configuration options.

### What issues does this PR fix or reference?
Fixes #52726

### Previous Behavior
The options mentioned aboved were ignored by libvirt provider, the cloned domain always inherited options from source domain.

### New Behavior
The options mentioned aboved are used by libvirt provider to customize cloned domains.

### Tests written?
No

### Commits signed with GPG?
No

### CODE REVIEW

**Comment by waynew** [comment]
Hey, thanks for the PR!

There's some issues with lint - mixed spaces and tabs apparently (should be all spaces).

Also, would it be possible to add some tests here?

**Comment by f-andriot** [comment]
Added more commits to add more features to libvirt provider.

Now you can:
- Define number of CPU
- Define amount of RAM
- Define several network connections
- Define guest serial number (as returned by dmidecode in guest)
- Create supplementary virtual disks

Also, fix issues during libvirt provisioning:
- Libvirt pool is not properly refreshed after VM provisioning (commit 
 #6c33758 )
- Do not attempt to connect to guest IP address beginning with '169' (invalid address given to unconfigured windows network interfaces) (commit #b1aeafb )

**Review by waynew — DISMISSED** [review]
Hey, this is looking great - just a couple of things:

1. We've switched to a `master` branch, so if you could change this PR to point to `master` and rebase your changes there that would be 🎉 
2. This code is lacking automated tests. To protect against regressions, code needs to have automated tests.

We  can help train you if you have desire to write tests but just lack the skill/knowledge.

Please let us know if you're interested in making those changes!

**Comment by f-andriot** [comment]
Hello, 
OK I will rebase on the master branch.
You are right, I have absolutely no knowledge about writing automatic tests. Please, do you have documentation about it ?

**Comment by waynew** [comment]
Unfortunately we don't yet have any good guides - if you go sign up for a time on here: https://calendar.google.com/calendar/selfsched?sstoken=UURWQ2dRZDlKU0twfGRlZmF1bHR8OWU3N2NmZjgwODY5ODVjNGU3NWRhMTRhYTI4MDM3ZmY I can meet with you either via chat or hangouts/zoom/???

**Comment by waynew** [comment]
@f-andriot we'll be running a test clinic this Thursday during our Community Open Hour. If you haven't subscribed to our Community Events Calendar you can get it [here](https://calendar.google.com/calendar/embed?src=saltstack.com_md73c3ufcs2eqbsmmnike4em80%40group.calendar.google.com). Click the *totally intuitive* + at the very bottom right of the screen to add it to your own calendar.

**Comment by waynew** [comment]
@f-andriot are you still interested in driving this to completion?

**Comment by f-andriot** [comment]
> @f-andriot are you still interested in driving this to completion?

Hello, I'm still working on developing for this PR inside my company, to add more features.
I will push again here when I'm done.
But I'll probably won't have time to make the automatic test stuff.

**Comment by waynew** [comment]
@f-andriot thanks for the update!

Let us know when you're done with the code side of things - we'll be hosting test clinics regularly, and if you're not interested or able to write the tests we'll find someone who will be able to write the tests 👍

**Comment by sagetherage** [comment]
@f-andriot please rebase this against the master branch and update the title with brackets if it is a WIP = Work In Progress on your end, we will help with tests, but if we have not heard from you in some time we will close this PR as abandoned.

**Comment by f-andriot** [comment]
Hello, I've rebased against master branch as requested.
But the PR is still against develop branch. Is there a way to change it ?

**Comment by sagetherage** [comment]
@f-andriot you can do a `git-rebase --onto` git-scm has some good reference guides [here](https://git-scm.com/docs/git-rebase) and looks like you have some merge conflicts, as well.

**Comment by f-andriot** [comment]
Hello, I've added new commits and rebased against master branch as requested.
Still, the PR was done agains the 'develop' branch, that's why it looks like there are many conflicting files.
Is there a way to change the destination branch ?

**Comment by waynew** [comment]
@f-andriot I went ahead and pointed the PR to master for you. If you have need of it in the future, all you have to do is click the "edit" at top right and then you can change where you're merging to 👍

**Comment by waynew** [comment]
re-run debian8-py3

**Comment by sagetherage** [comment]
@f-andriot Thank you for rebasing! looks like there is a merge conflict if you can resolve, please.

@cbosdo can you also give a review here, please?

File: salt/cloud/clouds/libvirt.py [hunk 1]
```diff
@@ -441,37 +594,203 @@ def create(vm_):
...
+
+                        if pool:
+                            pool_xml = ElementTree.fromstring(pool.XMLDesc())
+                            pool_target = pool_xml.find("./target/path").text
+
+                            vol_name = name + "-" + disk + "." + format
+                            vol_path = pool_target + "/" + vol_name
+
+                            try:
+                                vol = pool.storageVolLookupByName(vol_name)
```
Comment (cbosdo) [inline salt/cloud/clouds/libvirt.py:728]: You surely want to refresh the pool here. Maybe even ensure it is started...

**Review by cbosdo — CHANGES_REQUESTED** [review]
Thanks a lot for taking care of the libvirt cloud driver. Your PR seems fairly good to me, but I left a few comments to share code or make it more robust.

File: salt/cloud/clouds/libvirt.py [hunk 2]
```diff
@@ -390,11 +444,110 @@ def create(vm_):
...
+            # Configure number of CPU
+            if num_cpus:
+                vcpu_xml = domain_xml.find("./vcpu")
+                vcpu_xml.text = str(num_cpus)
+                log.debug("Setting CPU number to: %s", num_cpus)
+
+            # Configure serial number
+            if serial:
+                entry_xml = domain_xml.find("./sysinfo/system/entry")
+                entry_xml.text = str(serial)
```
[the remark this answers is not in the document]
Comment (f-andriot) [reply salt/cloud/clouds/libvirt.py:483]: OK I've added the code; the only problem is that "get_xml_node()" creates the XML tag at the end of XML file, whereas libvirt XML requires that <sysinfo> XML tag is at a specific location

File: salt/cloud/clouds/libvirt.py [hunk 3]
```diff
@@ -390,11 +444,110 @@ def create(vm_):
...
+                    elif memory_unit.lower() == "gb":
+                        memory_mb = int(float(memory_num) * 1024.0)
+                    else:
+                        err_msg = "Invalid memory type specified: '{}'".format(
+                            memory_unit
+                        )
+                        log.error(err_msg)
+                        return {"Error": err_msg}
+                except (TypeError, ValueError):
+                    memory_mb = int(memory)
```
[the remark this answers is not in the document]
Comment (f-andriot) [reply salt/cloud/clouds/libvirt.py:504]: This code is copy/paste from the vmware provider :)
From what I see in _handle_unit() , it is using per-1000 units, but we want per-1024 units here. E.g. user will ususally type "1GB" for virtual memory, and it should result in 2048MB, not 2000MB.

File: salt/cloud/clouds/libvirt.py (hunk 1 above)
Comment (f-andriot) [reply salt/cloud/clouds/libvirt.py:804]: OK, I've added the pool.refresh() and pool.create()

File: salt/cloud/clouds/libvirt.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbosdo) [reply salt/cloud/clouds/libvirt.py:483]: libvirt doesn't impose any order among sibling elements. You can perfectly add the `<sysinfo></sysinfo>` after `<devices></devices>`. Just mind that libvirt may change the place of the element though... so you may have a hard time diffing the `virsh dumpxml` output with what you provided libvirt.

File: salt/cloud/clouds/libvirt.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cbosdo) [reply salt/cloud/clouds/libvirt.py:504]: `_handle_unit` handles the units as defined by libvirt. So it handles both 1024 and 1000 versions. Typically `1MiB` is `1024KiB` while `1MB` is `1000KB`. Additionally `1m` will default to a 1024 multiple. This fits the usual megabyte vs [mebibyte definition](https://en.wikipedia.org/wiki/Byte#Multiple-byte_units)

File: salt/cloud/clouds/libvirt.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (f-andriot) [reply salt/cloud/clouds/libvirt.py:483]: Hello, that's what I though initially, but when I tried get_xml_node() directlry, it put the <sysinfo> XML tag at end of file as you describe, then I got a libvirt error regarding dictionary validation failure (in domaincommon.rng).
That's why I put the <sysinfo> on top of the<os> tag, now I have no dictionary error.
(BTW I'm using libvirt 4.5.0 from CentOS 7.9, not the latest libvirt version I guess)

File: salt/cloud/clouds/libvirt.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (f-andriot) [reply salt/cloud/clouds/libvirt.py:504]: OK I will use that, and I will add this precision to the comments too.
Older people (like me) still have 1MB = 1024KB in mind,  but this is wrong according to the definition.

[the remark this answers is not in the document]
Comment (f-andriot) [reply salt/cloud/clouds/libvirt.py]: Done in latest commit

File: salt/cloud/clouds/libvirt.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (f-andriot) [reply salt/cloud/clouds/libvirt.py:478]: OK this was a mistake in my code, after fixing my syntax it now works as expected. No need to create sysinfo at fixed location.

File: salt/cloud/clouds/libvirt.py [hunk 4]
```diff
@@ -66,6 +115,8 @@
     SaltCloudNotFound,
     SaltCloudSystemExit,
 )
+from salt.modules.virt import _handle_unit
```
Comment (cbosdo) [inline salt/cloud/clouds/libvirt.py:118]: I wonder if that function should be moved to `salt.utils.virt` module instead

File: salt/cloud/clouds/libvirt.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbosdo) [reply salt/cloud/clouds/libvirt.py:478]: cool, thanks!

File: salt/cloud/clouds/libvirt.py (hunk 1 above)
Comment (cbosdo) [reply salt/cloud/clouds/libvirt.py:781]: thanks

File: salt/cloud/clouds/libvirt.py [hunk 5]
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
...
+        _f for _f in [cfg.get("libvirt") for cfg in providers.values()] if _f
+    ]
+    for provider in providers_to_check:
+        conn = __get_conn(provider["url"])
+        log.info("looking at %s", provider["url"])
+        try:
+            domain = conn.lookupByName(name)
+            found.append(domain)
+        except libvirtError:
+            pass
```
Comment (cbosdo) [inline salt/cloud/clouds/libvirt.py:1275]: I would be better to close the connection in a `finally` block

File: salt/cloud/clouds/libvirt.py [hunk 6]
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
...
```
Comment (cbosdo) [inline salt/cloud/clouds/libvirt.py:1218]: Please close the connection in a `finally` block

File: salt/cloud/clouds/libvirt.py [hunk 7]
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
...
```
Comment (cbosdo) [inline salt/cloud/clouds/libvirt.py:1161]: Please close the connection in a `finally` block

File: salt/cloud/clouds/libvirt.py [hunk 8]
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
...
+        "event",
+        "starting instance",
+        "salt/cloud/{}/starting".format(name),
+        args={"name": name},
+        sock_dir=__opts__["sock_dir"],
+        transport=__opts__["transport"],
+    )
+
+    log.info("Starting domain %s", found[0].name())
+    found[0].create()
```
Comment (cbosdo) [inline salt/cloud/clouds/libvirt.py:1179]: You may get a `libvirtError` here too

File: salt/cloud/clouds/libvirt.py [hunk 9]
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
...
+        "event",
+        "stopping instance",
+        "salt/cloud/{}/stopping".format(name),
+        args={"name": name},
+        sock_dir=__opts__["sock_dir"],
+        transport=__opts__["transport"],
+    )
+
+    log.info("Stopping domain %s", found[0].name())
+    found[0].shutdown()
```
Comment (cbosdo) [inline salt/cloud/clouds/libvirt.py:1236]: You may also get a `libvirtError` here too.

File: salt/cloud/clouds/libvirt.py [hunk 10]
```diff
@@ -737,3 +1120,174 @@ def get_domain_volumes(conn, domain):
...
+        "event",
+        "rebooting instance",
+        "salt/cloud/{}/rebooting".format(name),
+        args={"name": name},
+        sock_dir=__opts__["sock_dir"],
+        transport=__opts__["transport"],
+    )
+
+    log.info("Rebooting domain %s", found[0].name())
+    found[0].shutdown()
```
Comment (cbosdo) [inline salt/cloud/clouds/libvirt.py:1293]: You may get a `libvirtError` here too

**Review by cbosdo — CHANGES_REQUESTED** [review]
There are still some libvirt connection handling improvements to be made, but the rest of the libvirt handling seems good.

I don't have sufficient salt-cloud knowledge to review the global design of your PR though. In my dreams this provider would use the `virt` module... but that is another story :wink:

File: salt/cloud/clouds/libvirt.py [hunk 11]
```diff
@@ -1159,6 +1159,8 @@ def start(name, call=None):
             found.append(domain)
         except libvirtError:
             pass
+        finally:
+            conn.close()
```
Comment (cbosdo) [inline salt/cloud/clouds/libvirt.py:1163]: Sorry for not having been precise enough. I think closing here may be too early: I would bet the domain objects stored in `found` array would then be freed.

There are two possible solutions:
1. Store the connection handles in a list and close them once all the domains have been started (or any other action for the similar functions)
2. Close the connections that gave no result within the `for` loop and the others with something like:
```python
for dom in found:
    dom.connect().close()
```

Comment (f-andriot) [reply salt/cloud/clouds/libvirt.py:1163]: Hello, I will look into that :)
I'm not the original author for these start, stop and reboot functions, so I need to talk with my colleague about what behaviour is expected exactly.

From what I understand, the functions are targeting a single VM, not a list.
To locate where the VM is, we connect to each provider in a loop.
If we find zero or more than one VM (e.g. there are VM with the same name on different hypervisor), we do nothing.

The functions are working in their current states, not sure how. I agree that closing the connection before calling start() should fail, but it does not.

**Comment by f-andriot** [comment]
OK so the start(), stop() and reboot() functions were copy/paste from existing destroy() function.
It looks like destroy() was already handling conn.close() , so I've done the same in our functions.

**Comment by Ch3LL** [comment]
ping @cbosdo is this one ready on your end?

File: salt/cloud/clouds/libvirt.py (hunk 11 above)
Comment (cbosdo) [reply salt/cloud/clouds/libvirt.py:1165]: OK, I see better how it works now. The real connection `close()` is done after the start so we are safe

**Review by cbosdo — APPROVED** [review]
LGTM

**Comment by Ch3LL** [comment]
Thanks for the review @cbosdo  Looks like the lint and pre-commit tests are failing and still need to be cleaned up @f-andriot

**Review by Ch3LL — CHANGES_REQUESTED** [review]
this will require a changelog and test coverage

**Comment by Ch3LL** [comment]
bump @f-andriot did you see my comments?

**Comment by dwoz** [comment]
Closing due to lack of activity. Anyone should feel free to re-open or create a new PR if they want to see this through to the end.

**Pull request closed without merging**